keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,232 @@
+
+High-level processes, such as starting a pod or restarting a failed container, could be interrupted before completion by an update to the desired state. While this leaves a "partial" trace for the first process, it is the most accurate representation of the work and timing of reconciling desired and actual state.
+
+#### End-User Behavior
+
+A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+
+Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+
+For example, a user could define a Pod Creation trace doing something like the following:
+```golang
+ctx, span := trace.StartSpan(context.Background(), “create-my-pod”)
+defer span.End()
+ctx = traceutil.WithSpanContext(ctx, span)
+pod, err := c.CoreV1().Pod(myPod.Namespace).Create(ctx, myPod)
+if err != nil {
+    return err
+}
+waitForPodToBeRunning(ctx, myPod)
+return nil
+```
+
+### In-tree changes
+
+#### Vendor the Tracing Framework
+
+This KEP proposes the use of the [OpenTelemetry tracing framework](https://opentelemetry.io/) to create and export spans to configured backends. However, this framework is still under development.  During the alpha phase, we will use [OpenCensus](https://opencensus.io/), which promises to be compatible with opentelemetry when it is released. 
+
+While in alpha, controllers should use the OpenCensus exporter, which exports traces to the [OpenCensus Agent](https://github.com/census-instrumentation/opencensus-service). The OpenCensus agent allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.
+
+This KEP suggests that we utilize the OpenCensus agent for the initial implementation to reduce the global changes required for alpha.  Alternative options include:
+
+1. Add configuration for exporters in-tree by vendoring in each "supported" exporter. 
```

Comment:
Can you explain what "supported" means here? It's not clear to me. I agree that supporting one exporter that can allow traces to be adapted out-of-process to the various backends is the way to go.

Reply:
It means a few things:

1. All "supported" exporters are compiled into all kubernetes binaries that support tracing.
2. Each component that supports tracing must add configuration options for each "supported" exporter.
3. An exporter that isn't "supported" cannot be used with kubernetes without recompiling binaries.